src/lib.rs

```diff
@@ -35,18 +35,23 @@ pub type SessionId = u64;
 
 /// The source of data items that consensus should order.
 ///
-/// AlephBFT internally calls [`DataIO::get_data`] whenever a new unit is created and data needs to be placed inside.
-/// The [`DataIO::send_ordered_batch`] method is called whenever a new round has been decided and thus a new batch of units
-/// (or more precisely the data they carry) is available.
+/// AlephBFT internally calls [`DataProvider::get_data`] whenever a new unit is created and data needs to be placed inside.
 ///
 /// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/aleph_bft_api.html for a discussion
 /// and examples of how this trait can be implemented.
-pub trait DataIO<Data> {
-    type Error: Debug + 'static;
+pub trait DataProvider<Data> {
     /// Outputs a new data item to be ordered
     fn get_data(&self) -> Data;
-    /// Takes a new ordered batch of data item.
-    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+/// The source of finalization of the units that consensus produces.
+///
+/// The [`FinalizationHandler::data_finalized`] method is called whenever a piece of data input to the algorithm
+/// using [`DataProvider::get_data`] has been finalized, in order of finalization.
+pub trait FinalizationHandler<Data> {
+    /// Data, provided by [DataProvider::get_data], has been finalized.
+    /// The calls to this function follow the order of finalization.
+    fn data_finalized(&self, data: Data);
```

Comment:
Consider changing to `&mut sefl` so at least tests and examples doesn't need to use `Arc/Mutex` or similar.

Reply:
Added